Esql/lookup join grammar

costin opened the pull request:
First PR for adding LOOKUP JOIN in ESQL.
Introduces grammar and wires the main building blocks to execute a query; follow-ups are required (see #116208 for more details).

costin committed:
first draft for grammar + lexer
rename existing lookup command to look_up (to avoid clashes)

costin committed:
Add first draft of parser and generated grammar

costin committed:
Finish up join field resolution

costin committed:
Change LOOK_UP to LOOKUP__ (to better signal the diff from JOIN)

costin committed:
Another set of polishing

nik9000 committed:
Hack mapping resolution for a specific lookup

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

nik9000 committed:
Plug?

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinTypes.java:
```diff
@@ -0,0 +1,145 @@
+     * Private class so it doesn't get used yet it is defined to showcase why the join type was defined as an interface instead of a simpler
+     * enum.
+     */
+    private abstract static class NaturalJoinType implements JoinType {
```
As `USING` establishes a natural  join, why not use/expand this "type"/class, instead of the `UsingJoinType`? Will we ever use them both?

costin replied:
Using and Natural are similar but different:
- in Using the keys are specified by the user. Using supports Inner, Left, right and full outer join.
- Natural - the keys are detected at runtime. Only supports inner join.

Hence why these are modelled differently.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinTypes.java:
```diff
@@ -0,0 +1,145 @@
+    /**
+     * The predefined core join types. Implements as enum for easy comparison and serialization.
+     */
+    private enum CoreJoinType implements JoinType {
```
This class implementing `JoinType` feels a bit odd -- does it need to? 
I _suppose_ it's so to be able to keep it private, so that the public constants can be used as is? But it feels a bit odd to init other implementations of `JoinType` (like  `UsingJoinType`) with a `JoinType`, to actually defines its behaviour.
In the Analyser we check for an instance of `UsingJoinType`, which is a `JoinType`, then get out of it the core type, which is another `JoinType`.
I guess what I'm wondering is if we could model the concepts of "how do we stitch" vs "what do we match" separately?

costin replied:
It's unusual - trying to mix the enum/limited functionality with extension for holding the keys specified by USING.
LEFT JOIN x USING a is different from LEFT JOIN x ON l.a == x.b (assuming namespace will be supported) - it's not just about the condition but also semantics (SQL USING returns the keys as the first column).

There's no strong reason to keep the core enum private; mainly to force the use of the interface and JoinTypes constants helps with evolution (before the enum was differently defined).

nik9000 committed:
Local is busted

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/mapper/MapperUtils.java:
```diff
@@ -53,7 +58,15 @@ private MapperUtils() {}
     static PhysicalPlan mapLeaf(LeafPlan p) {
         if (p instanceof Row row) {
-            return new RowExec(row.source(), row.fields());
+            // return new RowExec(row.source(), row.fields());
+            // convert row into local relation
+            List<Alias> fields = row.fields();
+            List<Object> values = new ArrayList<>(fields.size());
+            for (Alias field : fields) {
+                values.add(field.child().fold());
+            }
+            Block[] blocks = BlockUtils.fromListRow(PlannerUtils.NON_BREAKING_BLOCK_FACTORY, values);
+            p = new LocalRelation(row.source(), row.output(), LocalSupplier.of(blocks));
```
Makes sense. I guess we can purge `RowExec` entirely? (It shouldn't have ever been serialised.)
But was this intended in this PR?

costin replied:
It's a workaround to have the mapping works - otherwise it fails since it expects a local relationship and the row throws it off.
It's a topic we've identified before:
https://github.com/elastic/elasticsearch/issues/104960, https://github.com/elastic/elasticsearch/issues/106653

bpintea reviewed (commented):
Some minor questions.

costin committed:
Temporary disable FieldExtraction
Remove USING syntax - use ON <field> for now
Fix incorrect serialization

costin committed:
Merge branch 'main' into esql/lookup-join-grammar

costin committed:
disable docs for inlinestats and lookup

costin committed:
wip

costin committed:
Merge branch 'main' into esql/lookup-join-grammar

costin committed:
Add missing writable
Move writable declarations outside the core classes to avoid errors
(such as subClass.getNamedWritable()) and centralize them in a top
package class for better management.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java:
```diff
@@ -106,31 +101,49 @@ public List<Attribute> output() {
+    }
+
+    private static List<Attribute> removeDuplicateNames(List<Attribute> attributes, Set<String> sideNames, Set<String> matchFieldNames) {
+        List<Attribute> result = new ArrayList<>(attributes.size());
         for (Attribute attr : attributes) {
-            if ((matchFields.contains(attr) || matchFieldNames.contains(attr.name())) == false) {
+            String name = attr.name();
+            if ((sideNames.contains(name) || matchFieldNames.contains(name)) == false) {
                 result.add(attr);
             }
         }
```
This has become unused.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java:
```diff
@@ -106,31 +101,49 @@ public List<Attribute> output() {
+        }
+        for (Attribute a : right) {
+            // override the existing entry in place
+            nameToAttribute.compute(a.name(), (name, existing) -> a);
```
Guess we should make `a` nullable here.

costin replied:
Added a TODO plus entry in the meta ticket - making the attributes nullable changes their equality which has side effects in the plan.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java:
```diff
@@ -106,31 +101,49 @@ public List<Attribute> output() {
-        List<Attribute> result = new ArrayList<>();
+    /**
+     * Merge the two lists of attributes into one and preserves order.
+     * In case of conflicts, specify the existing entry is overridden (in place) or not.
```
Doc doesn't match implementation (yet?). Currently the right overrides the left, no way to specify otherwise.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinTypes.java:
```diff
@@ -0,0 +1,145 @@
+    public static JoinType readFrom(StreamInput in) throws IOException {
+        byte id = in.readByte();
+        return switch (id) {
+            case 1 -> INNER;
+            case 2 -> LEFT;
+            case 3 -> RIGHT;
+            case 4 -> FULL;
+            case 5 -> CROSS;
```
nit: maybe this can be made a little more robust by directly using the `CoreJoinType.id`s here, e.g. by iterating over the enum's values and throwing them into a map with ids (obtained from each enum element directly) as key. Would avoid repeating the assignment between enum element and id here.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/CsvTests.java:
```diff
@@ -170,7 +171,7 @@ public class CsvTests extends ESTestCase {
     @ParametersFactory(argumentFormatting = "%2$s.%3$s")
     public static List<Object[]> readScriptSpec() throws Exception {
-        List<URL> urls = classpathResources("/*.csv-spec");
+        List<URL> urls = classpathResources("/lookup-join.csv-spec");
```
Marking this for later as that's an important NOCOMMIT :)

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plan/logical/JoinSerializationTests.java:
```diff
@@ -27,7 +28,7 @@ protected Join createTestInstance() {
 
     private static JoinConfig randomJoinConfig() {
-        JoinType type = randomFrom(JoinType.values());
+        JoinType type = randomFrom(JoinTypes.LEFT, JoinTypes.RIGHT, JoinTypes.INNER);
```
Why do we restrict the join types in this serialization test?

costin replied:
Because natural and using don't survive analysis and thus don't/shouldn't get serialized.

alex-spies reviewed (commented):
I had a first pass with some questions and thoughts.

costin committed:
Remove NOCOMMIT

costin committed:
Test fixes

costin committed:
Formatting

fang-xing-esql reviewed (commented):
I added some thoughts on the grammar/parser first, will look into the rest next.

costin committed:
Fix test

costin committed:
🐔 FTW!

costin committed:
Merge branch 'main' into esql/lookup-join-grammar

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/lookup-join.csv-spec:
```diff
@@ -0,0 +1,42 @@
+| LIMIT 1;
+
+emp_no:integer | language_name:keyword
+;
```
Fixed

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4:
```diff
@@ -322,3 +323,19 @@ lookupCommand
+    ;
+
+joinTarget
+    : index=identifier (AS alias=identifier)?
```
I've added this to the metadata https://github.com/elastic/elasticsearch/issues/116208
It will be delicate since the command already uses identifiers which will clashes/is ambiguous with the indexSource declaration.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/UnaryScalarFunction.java:
```diff
@@ -75,62 +75,6 @@
 import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isNumeric;
 
 public abstract class UnaryScalarFunction extends EsqlScalarFunction {
-    public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
```
Should be fixed now - sorry for the noise.

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4:
```diff
@@ -322,3 +323,19 @@ lookupCommand
+    ;
+
+joinCondition
+    : ON joinPredicate (COMMA joinPredicate)*
```
ON a is the same as USING a.
Moving forward USING should be converted in an expression (after the field resolution happens since we don't have qualifiers and thus the name is not enough).

costin committed:
Address feedback

costin committed:
Update docs
Disable lookup join for local CSV tests
Enable back all csv tests
Disable failing test in CI
Update tests

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
Before merging, I believe we'll need a couple tests here. I can think of:
- LOOKUP JOIN up with an index that is not in lookup mode.
- LOOKUP JOIN with an index that does not exist.
- LOOKUP JOIN ON/USING a field that does not exist in the left/right hand side or both.
- LOOKUP JOIN with USING where there isn't a matching field in the left/right hand side because names or datatypes differ.
- LOOKUP JOIN ON/USING a field with a data type that we cannot join on - e.g. `_source`, spatial or whichever types we won't be able to join on in the beginning.
- Cases with currently disabled join types, like left/right/full

alex-spies replied:
On second thought, that's probably best addressed in a separate PR, but I think we should add it to the meta issue. I took the liberty of adding it to https://github.com/elastic/elasticsearch/issues/116208.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java:
Maybe not in this PR, but I believe we'll need some tests that ensure that lookup indices with all kinds of names, incl. special characters, and date math (good point @fang-xing-esql !) applied to them work correctly. Could also throw in `ON`/`USING` for subfields like `parent.subfield` _if_ we want to allow using such fields in lookup indices. And tests for whichever quoting convention we end up using.

costin replied:
Thanks, see top entry in #116208

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java:
```diff
@@ -1948,7 +1948,7 @@ public void testLookup() {
                  * Int is a reference here because we renamed it in project.
                  * If we hadn't it'd be a field and that'd be fine.
                  */
-                .item(containsString("int{r}"))
+                .item(containsString("int{f}"))
```
This is subtle. Marking this for before we merge: if this change remains in place, I think we should update the comment above to make sense. Or we just disable the test for now but leave the initial expectations in place.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -588,6 +620,76 @@ private LogicalPlan resolveLookup(Lookup l, List<Attribute> childrenOutput) {
             return l;
         }
 
+        private Join resolveLookupJoin(LookupJoin join) {
```
super duper nit:
```suggestion
        private LookupJoin resolveLookupJoin(LookupJoin join) {
```
I think it's relevant to convey that we're not mapping to a general join here - which would lose us information, rather than only adding information by means of resolving references.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -588,6 +620,76 @@ private LogicalPlan resolveLookup(Lookup l, List<Attribute> childrenOutput) {
+                List<Attribute> leftKeys = resolveUsingColumns(cols, join.left().output(), "left");
+                List<Attribute> rightKeys = resolveUsingColumns(cols, join.right().output(), "right");
+                List<Attribute> output = new ArrayList<>(join.left().output());
+                output.addAll(join.right().outputSet().subtract(new AttributeSet(rightKeys)));
```
This is subtle. Using the rhs `outputSet` is actually stable in terms of the order of the attributes, because AttributeMaps use a LinkedHashMap internally, right? Maybe we could point this out in a comment here. (Or we add a big, prominent note to AttributeMap's javadoc.)

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -588,6 +620,76 @@ private LogicalPlan resolveLookup(Lookup l, List<Attribute> childrenOutput) {
+                // update the config - pick the left keys as those in the output
+                type = new UsingJoinType(coreJoin, rightKeys);
+                config = new JoinConfig(type, leftKeys, leftKeys, rightKeys);
+                join = new LookupJoin(join.source(), join.left(), join.right(), config, output);
```
Why do we have to rewrite the config and the output here, in the first place? Shouldn't the output be computed from the left/right sides + the join config in `LookupJoin.output()`? Seems like we place the output computation into multiple places.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -588,6 +620,76 @@ private LogicalPlan resolveLookup(Lookup l, List<Attribute> childrenOutput) {
+                List<Attribute> leftKeys = resolveUsingColumns(cols, join.left().output(), "left");
+                List<Attribute> rightKeys = resolveUsingColumns(cols, join.right().output(), "right");
+                List<Attribute> output = new ArrayList<>(join.left().output());
+                output.addAll(join.right().outputSet().subtract(new AttributeSet(rightKeys)));
```
Is it really correct to subtract the right hand side's attributes based on name id, which we do here?

I'd expect the left and right hand side's attributes to _normally_ have completely disjoint output sets, esp. because we normally plan to join different indices. They _should_ not have field attributes with the same name ids; instead, we should match them up based on their names, no?

costin replied:
See #116208 "clarify output order" - tl;dr this is not final.
Longer:
you are right that each side, including common keys have different ids. the goal is to move away from name matching and convert the using into a equality condition which can then be pushed down and combined with per-side filtering.
That is, formalize the using into an equal-join that is a filter with equality filter which simplifies other optimizations such as transitive filtering (left_key < 1 AND left_key == right_key --> implies right_key < 1; the left_key filter and right_key filter end up pushed down on their respective sides).
This means moving away from JoinCondition to an expression

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -588,6 +620,76 @@ private LogicalPlan resolveLookup(Lookup l, List<Attribute> childrenOutput) {
+                    resolved.add(resolvedCol);
+                }
+                // columns are expected to be unresolved - if that's not the case return an error
+                else {
+                    return singletonList(
+                        new UnresolvedAttribute(
+                            col.source(),
+                            col.name(),
+                            "Surprised to discover column [ " + col.name() + "] already resolved"
+                        )
+                    );
+                }
```
I think this should be an IllegalStateException - otherwise we mask an internal server error (500) as an invalid query (400).

costin replied:
Added item to #116208

alex-spies reviewed (commented):
Time for another round!

I didn't review the contained refactor of the writables, as that's many changed lines and deserves its own careful review. I'd appreciate if we could do that in a separate PR.

alex-spies reviewed (changes requested):
Ok, all done now. I think we're not too far away, but I think there's a couple things that needs to be done before merging - and a bunch of things that can be done after.

Here's what I think we should do:
- I believe the refactoring of the named writable entries should become its own PR. Currently, our mixed cluster test suite is muted, which means that making such a large scale refactor to our serialization is quite risky.
- It looks to me like our current planning for lookup joins expects the match fields from the left and right hand side's outputs to _have the same name ids_. I believe this is wrong and will lead to inconsistent planning; see my comment on the changes to `Analyzer.java`. The fields in the lookup index are different from fields that we join on. Actually, even if we join the lookup index with itself, we should probably still assign different name ids to the respective field attrs. to avoid bugs. There's a comment in `LookupJoinExec` that claims we need qualifiers to make this work, but I don't believe so - name ids should disambiguate the attributes just fine.
- `LookupJoin` and `LookupJoinExec` have some inconsistencies and hacks which I think we best address to avoid them from lingering and producing secondary issues. E.g. the fact that `LookupJoinExec` doesn't have any right hand side attributes in its `references()`, the fact that `LookupJoin.info()` is not quite right etc. See the individual comments.
- Test coverage (can be addressed in a follow up, but then we need to put it in the meta issue as well):
  - The test for lookup joins on the data nodes doesn't look right to me. But we can consider lookup joins on data nodes a follow-up.
  - In general, the test coverage in terms of csv tests for lookup joins even on the coordinator is too small. We can address this in a follow-up, though, but at least should add it to the meta issue.

nik9000 committed:
Disable lookup in CCS

costin committed:
Update docs/changelog/116515.yaml

elasticsearchmachine commented:
Hi @costin, I've created a changelog YAML for you.

costin committed:
Merge branch 'main' into esql/lookup-join-grammar

costin commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/ExpressionCoreWritables.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ExpressionCoreWritables {
```
Indeed the reason for that was that each implementation subclass gained a `getNamedWritables()` so one (me :) ) that returned not the class writable but the entire hierarchy writables.
Which is confusing since Field.getNamedWritable() not only it's possible but doesn't return FieldAttribute.ENTRY but also MetadataAttribute and others. Plus it added a cyclic dependency.

I ended up working on it since the CsvTests had errors due to LookupJoinExec serialization not being written in and the serialization code style was split - see the change in EsqlPlugin.getNamedWritables (mix of utility classes plus individual entries).
I'll work look into extracting it in a separate PR.

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/lookup-join.csv-spec:
```diff
@@ -0,0 +1,48 @@
+FROM employees
+| EVAL language_code = languages
+| LOOKUP JOIN languages_lookup ON language_code
+| WHERE emp_no < 500
+| KEEP emp_no, language_name
+| SORT emp_no
+| LIMIT 1
```
The tests didn't start out this way however it triggered a number of issues including with MixedCluster tests (related to CCS) - see #116208.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/LookupJoin.java:
```diff
@@ -0,0 +1,84 @@
+        return new Project(source(), normalized, output);
+    }
+
+    public List<Attribute> output() {
+        return output;
+    }
```
I've added a line in #116208.
The main reason why right now it's in the Analyzer is because in case of failure, the LookupJoin becomes an unresolved Join since it doesn't survive (it gets rewritten to a regular join with a special output due to the particularities of using).
Separately the logic around LEFT/RIGHT, etc.. types of join and dealing with each branch are harder to encapsulate because the class has to check itself whether it's children are resolved, what's left and right side.
This type of pattern matching and traversal is done outside the tree classes.

See the TODO I added in JoinConfig and my previous comments.
JoinConfig models an equijoin however moving forward a condition expression since it encapsulates better the generality in the future (ON a > b) and simplifies tree traversal (JoinConfig doesn't extend Node and thus cannot be visited by our rules).

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/LookupJoin.java:
```diff
@@ -0,0 +1,84 @@
+
+    @Override
+    protected NodeInfo<Join> info() {
+        // Do not just add the JoinConfig as a whole - this would prevent correctly registering the
+        // expressions and references.
+        return NodeInfo.create(this, LookupJoin::new, left(), right(), config(), output);
```
The comment was no longer relevant and was removed.

alex-spies replied:
I think we still have to pass the individual components of the join config to the node info instead of the join config object as a whole.

If we want to, say, rewrite the plan and need to rename an attribute in the join config's left hand side, just calling `forEachExpression` on the join will be wrong because it ignores the join config.

We had issues because of this with `Join` before; let's fix it so we don't run into trouble later.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/LookupJoin.java:
```diff
@@ -0,0 +1,84 @@
+
+/**
+ * Lookup join - specialized LEFT (OUTER) JOIN between the main left side and a lookup index (index_mode = lookup) on the right.
+ * In the future, as the join capabilities of the engine will evolve, a regular LEFT JOIN will be used instead, letting the planner decide
+ * the strategy at runtime.
```
Thanks, updated the javadoc.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/mapper/Mapper.java:
```diff
@@ -202,6 +204,19 @@ private PhysicalPlan mapBinary(BinaryPlan bp) {
                     join.output()
                 );
             }
+            if (right instanceof FragmentExec fragment
+                && fragment.fragment() instanceof EsRelation relation
+                && relation.indexMode() == IndexMode.LOOKUP) {
+                return new LookupJoinExec(
+                    join.source(),
+                    left,
+                    right,
+                    config.matchFields(),
+                    config.leftFields(),
+                    config.rightFields(),
+                    join.output()
+                );
```
I agree right now the abstraction is simplistic but the point is to decouple the command from the execution.

LookupJoin is syntactic sugar for special left join with special output - it gets folded into a regular join since I'd like to avoid serializing it (and dealing with versioning) and to have the compute engine pick up the pattern not the class.
That means any syntax that gets converted into a left join could be executed against a lookup join. Enrich for example could be attached onto this one but the dependency between the logical and physical happens through the pattern NOT the Lookup class.

alex-spies replied:
That's fair - but we aren't at the point yet where an Enrich, or anything else for that matter, is rewritten into a pattern that is translated into a LookupJoinExec. Currently, we only have the Lookup logical plan node, which doesn't need the additional round trip via surrogate + pattern matching.

I really prefer deferring such generalizations until they have actual use cases - because getting the generalization wrong early on leads to pain :/ and it makes the code more complex than it currently needs.

But that's supposed to be a nit anyway :)

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/LookupJoinExec.java:
```diff
@@ -0,0 +1,160 @@
+
+    @Override
+    public AttributeSet inputSet() {
+        // TODO: this is a hack until qualifiers land since the right side is always materialized
+        return left().outputSet();
+    }
```
I've updated the comment.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/LookupJoinExec.java:
```diff
@@ -0,0 +1,160 @@
+
+    @Override
+    protected AttributeSet computeReferences() {
+        return Expressions.references(leftFields);
```
added a TODO

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/lookup-join.csv-spec:
Thanks for adding these to #116208

costin commented:
Thanks @alex-spies for all the comments, see my nested replies. This PR doesn't aim to fully address the lookup join, rather add the basic layers that can later be distributed and worked on as a follow-up from https://github.com/elastic/elasticsearch/issues/116208 in which you already commented (thanks for that!).

costin committed:
Address feedback

alex-spies reviewed (approved):
Heya - approving this to unblock the PR. I think it's fine to address most things in follow ups, as you suggested @costin.

Before merging, I think it'd be great to
- merge https://github.com/elastic/elasticsearch/pull/117029 first so that the serialization refactoring can be easily backported,
- carefully consider if this should be backported as well (not doing so can seriously hamper our ability to auto-backport as this PR touches many places, even when factoring out the serialization changes), and
- fix `LookupJoin.info()` as commented before.

costin committed:
Merge branch 'main' into esql/lookup-join-grammar
